File: python/백준/자료구조/스택, 큐, 덱/17298 오큰수.py
n = int(input())

array = list(map(int, input().split()))
result = [-1] * n
stack = []

stack.append(0)

for i in range(1, n):
    while stack and array[stack[-1]] < array[i]:
        result[stack.pop()] = array[i]
        
    stack.append(i)
    
    
print(*result)z

File: cpp/백준/자료구조/스택, 집합/17298 오큰수.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<int> array(n);
    for (int i = 0; i < n; ++i) cin >> array[i];

    vector<int> result(n, -1);
    stack<int> st;
    st.push(0);

    for (int i = 1; i < n; ++i) {
        while (!st.empty() && array[st.top()] < array[i]) {
            result[st.top()] = array[i];
            st.pop();
        }
        st.push(i);
    }

    for (int i = 0; i < n; ++i) {
        cout << result[i];
        if (i + 1 < n) cout << ' ';
    }
    cout << "\n";
    return 0;
}
